File: backend/app/schemas/news.py
from datetime import datetime
from typing import Optional
from pydantic import BaseModel


class NewsPostCreate(BaseModel):
    title: str
    content: str
    city: Optional[str] = None


class NewsAuthor(BaseModel):
    id: int
    first_name: str
    last_name: str
    avatar_url: Optional[str] = None

    model_config = {"from_attributes": True}


class NewsImageOut(BaseModel):
    id: int
    image_url: str
    order: int

    model_config = {"from_attributes": True}


class NewsPostOut(BaseModel):
    id: int
    title: str
    content: str
    image_url: Optional[str] = None
    images: list[NewsImageOut] = []
    city: Optional[str] = None
    author: Optional[NewsAuthor] = None
    event_id: Optional[int] = None
    event_image_url: Optional[str] = None
    created_at: datetime

    model_config = {"from_attributes": True}

    @classmethod
    def from_post(cls, post) -> "NewsPostOut":
        data = cls.model_validate(post)
        if not data.image_url and not data.images and post.event_id and post.event:
            data.event_image_url = post.event.image_url
        return data

File: frontend/src/api/news.ts
import api from './client'

export interface NewsAuthor {
  id: number
  first_name: string
  last_name: string
  avatar_url?: string
}

export interface NewsImage {
  id: number
  image_url: string
  order: number
}

export interface NewsPost {
  id: number
  title: string
  content: string
  image_url?: string
  images: NewsImage[]
  city?: string
  author: NewsAuthor
  event_id?: number
  event_image_url?: string
  created_at: string
}

export const newsApi = {
  list: (city?: string) =>
    api.get<NewsPost[]>('/news', { params: city ? { city } : {} }),

  create: (data: FormData) =>
    api.post<NewsPost>('/news', data, {
      headers: { 'Content-Type': 'multipart/form-data' },
    }),

  uploadImages: (postId: number, files: File[]) => {
    const form = new FormData()
    files.forEach((f) => form.append('files', f))
    return api.post<NewsPost>(`/news/${postId}/images`, form)
  },

  deleteImage: (postId: number, imageId: number) =>
    api.delete(`/news/${postId}/images/${imageId}`),

  delete: (id: number) =>
    api.delete(`/news/${id}`),
}
